asserts/hardware_identity_test.go

```diff
@@ -138,3 +144,232 @@ func (s *hardwareIdentitySuite) TestDecodeInvalid(c *C) {
+	pubKey, privKey, err := ed25519.GenerateKey(rand.Reader)
+	c.Assert(err, IsNil)
+
+	h, err := buildHardwareIdentityAssertion(pubKey)
+	c.Assert(err, IsNil)
+
+	nonce := []byte("test nonce")
+	hash := sha3.New384()
+	hash.Write(nonce)
+	hashed := hash.Sum(nil)
+
+	signature := ed25519.Sign(privKey, hashed)
+
+	err = h.VerifyNonceSignature(nonce, signature, crypto.SHA3_384)
+	c.Assert(err, IsNil)
+
+	err = h.VerifyNonceSignature(nonce, append(signature, 1), crypto.SHA3_384)
+	c.Assert(err, ErrorMatches, "signature invalid")
+}
+
+func (s *hardwareIdentitySuite) TestVerifySignatureDifferentHashAlgorithm(c *C) {
+	privKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	c.Assert(err, IsNil)
+
+	h, err := buildHardwareIdentityAssertion(&privKey.PublicKey)
+	c.Assert(err, IsNil)
+
+	nonce := []byte("test nonce")
+	// Sign with SHA256
+	hash := sha256.New()
+	hash.Write(nonce)
+	hashed := hash.Sum(nil)
+
+	signature, err := rsa.SignPKCS1v15(rand.Reader, privKey, crypto.SHA256, hashed)
+	c.Assert(err, IsNil)
+
+	// Verify with correct hash algorithm - should pass
+	err = h.VerifyNonceSignature(nonce, signature, crypto.SHA256)
+	c.Assert(err, IsNil)
+
+	// Verify with different hash algorithm (SHA2-512) - should fail
+	err = h.VerifyNonceSignature(nonce, signature, crypto.SHA512)
+	c.Assert(err, ErrorMatches, ".*verification error")
+
+	// Sign with SHA2-512 and verify with SHA2-512 - should pass
+	hash512 := crypto.SHA512.New()
+	hash512.Write(nonce)
+	hashed512 := hash512.Sum(nil)
+
+	signature512, err := rsa.SignPKCS1v15(rand.Reader, privKey, crypto.SHA512, hashed512)
+	c.Assert(err, IsNil)
+
+	err = h.VerifyNonceSignature(nonce, signature512, crypto.SHA512)
+	c.Assert(err, IsNil)
+}
+
+func (s *hardwareIdentitySuite) TestVerifySignatureUnsupportedHashOrAlgorithm(c *C) {
+	privKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	c.Assert(err, IsNil)
+
+	h, err := buildHardwareIdentityAssertion(&privKey.PublicKey)
+	c.Assert(err, IsNil)
+
+	const UNSUPPORTED_HASH = crypto.Hash(0) // Invalid hash value
+	err = h.VerifyNonceSignature(nil, nil, UNSUPPORTED_HASH)
+	c.Assert(err, ErrorMatches, "unsupported hash type: .*")
+
+	var params dsa.Parameters
+	err = dsa.GenerateParameters(&params, rand.Reader, dsa.L1024N160)
+	c.Assert(err, IsNil)
+
+	privDSAKey := new(dsa.PrivateKey)
+	privDSAKey.Parameters = params
+	err = dsa.GenerateKey(privDSAKey, rand.Reader)
+	c.Assert(err, IsNil)
+
+	h, err = buildHardwareIdentityAssertion(&privDSAKey.PublicKey)
+	c.Assert(err, IsNil)
+
+	err = h.VerifyNonceSignature(nil, nil, crypto.SHA256)
+	c.Assert(err, ErrorMatches, "unsupported algorithm type: .*")
+}
+
+func buildHardwareIdentityAssertion(hardwareKey crypto.PublicKey) (*asserts.HardwareIdentity, error) {
+	var pubBytes []byte
+	var err error
+
+	// DSA keys need special handling since x509.MarshalPKIXPublicKey doesn't support them
+	if dsaKey, ok := hardwareKey.(*dsa.PublicKey); ok {
+		pubBytes, err = marshalDSAPublicKey(dsaKey)
```

Comment:
do I understand correctly that Marshal doesn't support DSA but ParsePKIXPublicKey doesn't?